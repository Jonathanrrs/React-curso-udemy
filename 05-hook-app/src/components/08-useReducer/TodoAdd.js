import React from 'react'
import { UseForm } from '../../hooks/UseForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [{ description }, handleInputChange, reset] = UseForm({
        description: '', /* este viene del formulario de abajo */
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }


        handleAddTodo(newTodo);
        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    onChange={handleInputChange}
                    value={description}
                />
                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
