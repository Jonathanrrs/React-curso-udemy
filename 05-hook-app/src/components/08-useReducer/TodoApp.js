import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import './styles.css';
import { UseForm } from '../../hooks/UseForm';


// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'Aprender React',
//     done: false
// }];

// const init = () => {
//     return [{
//         id: new Date().getTime(),
//         desc: 'Aprender React',
//         done: false
//     }];
// }

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
}

export const TodoApp = () => {


    const [todos, dispatch] = useReducer(todoReducer, [], init); /* por defecto es un arreglo vacio */

    const [{description}, handleInputChange, reset] = UseForm({
        description: '', /* este viene del formulario de abajo */
    });

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (todoId) => {
        console.log(todoId);
    }


    console.log(description);

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        const action = {
            type: 'add',
            payload: newTodo
        }

        dispatch(action);
        reset();
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, i) => (
                                <li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p className="text-center complete">{i + 1}. {todo.desc}</p>
                                    <button
                                        className="btn btn-danger"
                                    >
                                        Borrar
                        </button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
                <div className="col-5">
                        <h4>Agregar TODO</h4>
                        <hr />
                        <form onSubmit={handleSubmit}>
                            <input 
                                type="text"
                                name="description"
                                className="form-control"
                                placeholder="Aprender..."
                                autoComplete="off"
                                onChange={handleInputChange}
                                value={description}
                            />
                            <button
                                type="submit"
                                className="btn btn-outline-primary mt-1 btn-block"
                                onClick={handleDelete}
                            >
                                Agregar
                            </button>
                        </form>
                </div>
            </div>
        </div>

    )
}
