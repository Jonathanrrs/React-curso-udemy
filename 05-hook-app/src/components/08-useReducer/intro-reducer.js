const initialState = [{
    id:1,
    todo: 'Comparar leche',
    done: false
}];



const todoReducer = (state = initialState, action) => {

    if(action?.type === 'agregar') {
        return [...state, action.payload]
    }
    return state;
}

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comparar pan',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    // newTodo: newTodo
    payload: newTodo
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos);