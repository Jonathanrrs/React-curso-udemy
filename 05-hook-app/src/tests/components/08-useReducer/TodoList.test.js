import React from 'react';
import {shallow} from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fuxtures/demoTodos';

describe('Pruebas en el componente', () => {
   
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
   
    const wrapper = shallow(<TodoList 
        todos = {demoTodos}
        handleDelete ={handleDelete}
        handleToggle ={handleToggle}
    />)
    
    test('Debe mostrarse correctamente ', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('Debe tener dos TodoListItem ', () => {
       
        expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
        /* se espera cualquier funcion, debe ser con toEqual porque con el tobe se intenta comparar con el valor de equidad 2 objetos diferentes, aunque sea igual como son objetos apuntan a espacios diferentes de memoria */
        expect(wrapper.find('TodoListItem').at(0).props('handleDelete')).toEqual(expect.any(Function));
    });
    
    
})
