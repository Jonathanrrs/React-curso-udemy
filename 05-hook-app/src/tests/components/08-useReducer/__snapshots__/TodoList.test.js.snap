// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente Debe mostrarse correctamente  1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender GO",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender python",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
